docs: initial document for comparing web embeds (#20478)
electron/electron: docs/tutorial/web-embeds.md

# Web embeds in Electron

> If you want to show (third party) web content in Electron, there are three options available to you: iframes, webviews and browserviews. Each one offers slightly different functionality and is useful in different situations. To help you choose between these this guide will explain the differences and capabilities of each.

Reviewer: ```suggestion
If you want to embed (third party) web content in an Electron `BrowserWindow`, there are three options available to you: `<iframe>` tags, `<webview>` tags, and `BrowserViews`. Each one offers slightly different functionality and is useful in different situations. To help you choose between these, this guide will explain the differences and capabilities of each.
```

## Iframes

> Iframes in Electron behave like iframes in regular browsers. An <iframe> element in your page can show external web pages, provided that their [Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) allows it. To limit the amount of capabilities a site in an iframe, it's recommended to use the [`sandbox` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#attr-sandbox) and only allow the capabilities you want to support.

Reviewer: ```suggestion
Iframes in Electron behave like iframes in regular browsers. An `<iframe>` element in your page can show external web pages, provided that their [Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) allows it. To limit the amount of capabilities a site in an `<iframe>` tag, it's recommended to use the [`sandbox` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#attr-sandbox) and only allow the capabilities you want to support.
```

> ## Webviews

Reviewer 2: ```suggestion
## WebViews
```

Author: We don't name them "WebViews" anywhere (actually we only name them lowercased: https://electronjs.org/docs/api/webview-tag but that felt weird for a heading)

Author: Oh, I see I do that in the line below. :sweat_smile:  I'll commit this :)

> [Webviews](https://electronjs.org/docs/api/webview-tag) are based on Chromium's webviews and are not explicitly supported by Electron, in such way that we do not give guarantees that the Webview api will remain available in future versions of Electron. This is why, if you want to use webviews, you will need to set `webviewTag` to `true` in the `webPreferences` of your BrowserWindow.

Reviewer: ```suggestion
[WebViews](https://electronjs.org/docs/api/webview-tag) are based on Chromium's WebViews and are not explicitly supported by Electron. We do not give guarantee that the WebView API will remain available in future versions of Electron. This is why, if you want to use `<webview>` tags, you will need to set `webviewTag` to `true` in the `webPreferences` of your BrowserWindow.
```

> [WebViews](../api/webview-tag) are based on Chromium's WebViews and are not explicitly supported by Electron. We do not guarantee that the WebView API will remain available in future versions of Electron. This is why, if you want to use `<webview>` tags, you will need to set `webviewTag` to `true` in the `webPreferences` of your `BrowserWindow`.

Reviewer: You got to use the "full" link to make `lint` pass:

```suggestion
[WebViews](../api/webview-tag.md) are based on Chromium's WebViews and are not explicitly supported by Electron. We do not guarantee that the WebView API will remain available in future versions of Electron. This is why, if you want to use `<webview>` tags, you will need to set `webviewTag` to `true` in the `webPreferences` of your `BrowserWindow`.
```

> [WebViews](https://electronjs.org/docs/api/webview-tag) are based on Chromium's WebViews and are not explicitly supported by Electron. We do not guarantee that the WebView API will remain available in future versions of Electron. This is why, if you want to use `<webview>` tags, you will need to set `webviewTag` to `true` in the `webPreferences` of your BrowserWindow.

Reviewer 2: ```suggestion
[WebViews](../api/webview-tag) are based on Chromium's WebViews and are not explicitly supported by Electron. We do not guarantee that the WebView API will remain available in future versions of Electron. This is why, if you want to use `<webview>` tags, you will need to set `webviewTag` to `true` in the `webPreferences` of your `BrowserWindow`.
```

> Webviews are a custom element(<webview>) that will only work inside Electron.

Reviewer: ```suggestion
WebViews are a custom element (`<webview>`) that will only work inside Electron.
```

> They are implemented as an "out-of-process iframe". This means all communication with the webview is done asynchronously using IPC. The webview element has many custom methods and events, similar to webContents, that allow you much greater control over the contents of a webview.

Reviewer: ```suggestion
They are implemented as an "out-of-process iframe". This means that all communication with the `<webview>` is done asynchronously using IPC. The `<webview>` element has many custom methods and events, similar to `webContents`, that allow you much greater control over the contents.
```

> Compared to an iframe, webviews tend to be slightly slower, but offer much greater control in loading and communicating with the third party content and handling various events.

Reviewer: ```suggestion
Compared to an `<iframe>`, `<webview>` tend to be slightly slower but offer much greater control in loading and communicating with the third party content and handling various events.
```

> ## Browserviews

Reviewer: ```suggestion
## BrowserViews
```

> Compared to an `<iframe>`, `<webview>` tend to be slightly slower but offer much greater control in loading and communicating with the third party content and handling various events.

Reviewer 2: ```suggestion
Compared to an `<iframe>`, `<webview>` tends to be slightly slower but offers much greater control in loading and communicating with the third party content and handling various events.
```

> Browserviews are not part of your DOM, but are created in your main process and are overlaid on top of your BrowserWindow content. This means they are completely separate from your own BrowserWindow content and their position is not controlled by the DOM and CSS, but by setting the bounds in the main process.

Reviewer: ```suggestion
Browserviews are not part of the DOM - instead, they are created in and controlled by your main process. They are simply another layer of web content on top of your existing window. This means that they are completely separate from your own BrowserWindow content and that their position is not controlled by the DOM or CSS but by setting the bounds in the main process.
```

> Browserviews are not part of the DOM - instead, they are created in and controlled by your main process. They are simply another layer of web content on top of your existing window. This means that they are completely separate from your own BrowserWindow content and that their position is not controlled by the DOM or CSS but by setting the bounds in the main process.

Reviewer 2: ```suggestion
[BrowserViews](../api/browser-view) are not part of the DOM - instead, they are created in and controlled by your main process. They are simply another layer of web content on top of your existing window. This means that they are completely separate from your own `BrowserWindow` content and that their position is not controlled by the DOM or CSS but by setting the bounds in the main process.
```

> [BrowserViews](../api/browser-view) are not part of the DOM - instead, they are created in and controlled by your main process. They are simply another layer of web content on top of your existing window. This means that they are completely separate from your own `BrowserWindow` content and that their position is not controlled by the DOM or CSS but by setting the bounds in the main process.

Reviewer: You got to use the "full" link to make `lint` pass:

```suggestion
[BrowserViews](../api/browser-view.md) are not part of the DOM - instead, they are created in and controlled by your main process. They are simply another layer of web content on top of your existing window. This means that they are completely separate from your own `BrowserWindow` content and that their position is not controlled by the DOM or CSS but by setting the bounds in the main process.
```

> BrowserViews offer the greatest control over their contents, since they implement the `webContents` similarly to how a BrowserWindow implements it. However they are not part of your DOM but are overlaid on top of them, which means you will have to manage their position manually.

Reviewer 2: ```suggestion
BrowserViews offer the greatest control over their contents, since they implement the `webContents` similarly to how a `BrowserWindow` implements it. However, they are not part of your DOM but are overlaid on top of them, which means you will have to manage their position manually.
```
